plugins/jobtype/test/azkaban/jobtype/TestTuningParameterUtils.java

```diff
@@ -0,0 +1,164 @@
+
+    Map<String, String> confProperties = TuningParameterUtils.getHadoopProperties(props);
+    Assert.assertTrue("Failed to get hadoop properties " + MRJobConfig.IO_SORT_MB,
+        confProperties.containsKey(MRJobConfig.IO_SORT_MB));
+    Assert.assertTrue("Failed to get right value for hadoop properties " + MRJobConfig.IO_SORT_MB,
+        confProperties.get(MRJobConfig.IO_SORT_MB).equals("100"));
+
+    Assert.assertTrue("Failed to get hadoop properties " + MRJobConfig.IO_SORT_FACTOR,
+        confProperties.containsKey(MRJobConfig.IO_SORT_FACTOR));
+    Assert.assertTrue("Failed to get right value for hadoop properties " + MRJobConfig.IO_SORT_FACTOR, confProperties
+        .get(MRJobConfig.IO_SORT_FACTOR).equals("100"));
+
+    Assert.assertTrue("Failed to get hadoop properties " + MRJobConfig.REDUCE_MEMORY_MB,
+        confProperties.containsKey(MRJobConfig.REDUCE_MEMORY_MB));
+    Assert.assertTrue("Failed to get right value for hadoop properties " + MRJobConfig.REDUCE_MEMORY_MB, confProperties
+        .get(MRJobConfig.REDUCE_MEMORY_MB).equals("2048"));
+
+    Assert.assertTrue("Failed to get hadoop properties " + MRJobConfig.MAP_MEMORY_MB,
+        confProperties.containsKey(MRJobConfig.MAP_MEMORY_MB));
+    Assert.assertTrue("Failed to get right value for hadoop properties " + MRJobConfig.MAP_MEMORY_MB, confProperties
+        .get(MRJobConfig.MAP_MEMORY_MB).equals("2048"));
+  }
+
+  @Test
+  public void testGetDefaultJobParameterJSON() throws IOException {
+    Props props = new Props();
+    props.put(HadoopConfigurationInjector.INJECT_PREFIX + MRJobConfig.IO_SORT_MB, "100");
+    props.put(HadoopConfigurationInjector.INJECT_PREFIX + MRJobConfig.IO_SORT_FACTOR, "100");
+    props.put(HadoopConfigurationInjector.INJECT_PREFIX + MRJobConfig.REDUCE_MEMORY_MB, "2048");
+    props.put(HadoopConfigurationInjector.INJECT_PREFIX + MRJobConfig.MAP_MEMORY_MB, "2048");
+
+    String defaultParamJSON = TuningParameterUtils.getDefaultJobParameterJSON(props);
+
+    String expectedJSON =
+        "{\"" + MRJobConfig.IO_SORT_MB + "\":\"100\"," + "\"" + MRJobConfig.MAP_MEMORY_MB + "\":\"2048\"," + "\""
+            + MRJobConfig.REDUCE_MEMORY_MB + "\":\"2048\"," + "\"" + MRJobConfig.IO_SORT_FACTOR + "\":\"100\"}";
+
+    Assert.assertEquals("Wrong JSON return for default job parameter ", defaultParamJSON, expectedJSON);
+  }
+
+  @Test
+  public void testUpdateAutoTuningParameters() throws ClientProtocolException, IOException {
+
+    String expectedJSON =
+        "{\"" + MRJobConfig.IO_SORT_MB + "\":\"300\"," + "\"" + MRJobConfig.MAP_MEMORY_MB + "\":\"1024\"," + "\""
+            + MRJobConfig.REDUCE_MEMORY_MB + "\":\"1024\"," + "\"" + MRJobConfig.IO_SORT_FACTOR + "\":\"10\"}";
+    PowerMockito.mockStatic(HttpClients.class);
+    PowerMockito.mockStatic(EntityUtils.class);
+
+    CloseableHttpClient httpClient = Mockito.mock(CloseableHttpClient.class);
+    PowerMockito.when(HttpClients.createDefault()).thenReturn(httpClient);
```

Comment:
Though `PowerMockito`'s static mocking abilities are definitely powerful, it also relies on rewriting bytecode to redirect calls somewhat magically. My inclination is that this should only be used as a last resort, for example when dealing with library code that you cannot easily modify. I would prefer to see the `updateAutoTuningParameters()` method refactored slightly to allow for you to pass in your own `HttpClient`, avoiding this type of mocking. Let me know if you disagree.

Reply:
Agree. Removing PowerMockito's usage. Changed TuningParameterUtils to have all methods as static.